apis/iam/v1beta1/rolepolicy_types.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2023 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// RolePolicyParameters define the desired state of an AWS IAM Role Inline Policy.
+type RolePolicyParameters struct {
+
+	// The JSON policy document that is the content for the policy.
+	Document string `json:"document"`
```

Comment:
Misformatted raw JSON policies are a frequent cause for errors. Could this be change to a `ResourcePolicy` as using in https://github.com/crossplane-contrib/provider-aws/blob/master/apis/kafka/v1alpha1/custom_types.go#L37?

That would give users a structured schema and avoids reconciler errors due to a missing comma. It is also much easier to handle in compositions.

Reply:
@MisterMX - fair point on the compositions
I just took a look at moving from `string` to `ResourcePolicy` and tested out that config.
We actually prefer strings instead.   I would guess since most users of this are using crossplane provider-aws IAM Policy in string format that moving from Policy + RolePolicyAttachment to this new RolePolicy would prefer as well and make this an easier transition.

Our use case is we deploy a LOT of IAM roles for 3rd party applications - probably more IAM for 3rd party applications than for our own webservices.  Those 3rd parties normally provide their IAM policies in json format.   The challenge with using resource policy is that we have to then convert their policy from a string which is natively supported by crossplane into the schema of the `ResourcePolicy`.  We then have to diff and maintain whenever those 3rd party roles change, convert them to the new schema.  There is more overhead required in doing that vs just dropping in a new json file.

For reference see ->  https://github.com/sergkondr/karpenter/blob/main/website/content/en/preview/getting-started/getting-started-with-karpenter/cloudformation.yaml#L41-L221
It would be a pain to convert that to `ResourcePolicy` and then have to maintain that

What we do instead is we just drop a json file in a helm `files/` directory and then convert it to json and then back to a string which allows us to maintain policies in the native 3rd party provided json format, and then use helm tooling to both customize and also verify that it is valid json.   This has caught many policy formatting issues.
Example:
```
---
apiVersion: iam.aws.crossplane.io/v1beta1
kind: Policy
metadata:
  name: aws-karpenter-{{ .Values.cluster }}
spec:
  forProvider:
    name: aws-karpenter-{{ .Values.cluster }}
    document: |
    {{- .Files.Get "files/aws-karpenter.json" | 
              replace `${AWS_PARTITION}` .Values.aws.partition | 
              replace `${AWS_REGION}` .Values.aws.region  | 
              replace `${CLUSTER_NAME}` .Values.cluster | 
              replace `${AWS_ACCOUNT_ID}` .Values.aws.account_id | 
              fromJson | mustToPrettyJson |
              nindent 6 }}
    description: aws-karpenter policy created by crossplane
    tags:
      - key: Environment
        value: {{ .Values.environment }}
  providerConfigRef:
    name: {{ .Values.providerConfigRef }}
```

You bring up a good point about the compositions and there are people who may prefer `ResourcePolicy`
I attempted to see if I could convert a json policy to the schema required by ResourcePolicy - I am still working on this - it's a challenge because the schema of `ResourcePolicy` differs so much from aws policy.

What are your thoughts with starting out just supporting string?  Or supporting both `string` and `ResourcePolicy` in the future?  Or maybe there are other solutions which would make sense?  One that I can think of which we might actually like is being able to `Ref` to the policy - and that Ref could be a configmap with policy json or a custom object which contains a `ResourcePolicy` object - just trying to think out of the box on something that could meet everyone's needs.

For now, for us, string is the easiest to support and maintain.

Reply:
I understand what you mean. Indeed, `ResourcePolicy` is a bit tricky to handle.

However, having a raw JSON strings can also be difficult since its schema is only validated at runtime. It is very easy to have misformatted JSON that causes reconciling to fail.

Do you think [`JSON`](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1#JSON) could be an alternative? It is used already in other places as well as https://github.com/crossplane/crossplane.

Reply:
@MisterMX when I change the type to
```
	// The JSON policy document that is the content for the policy.
	Document v1.JSON `json:"document"`
```
the CRD which is generated is broken
![image](https://github.com/crossplane-contrib/provider-aws/assets/14805373/59dfc0a1-7e86-40f1-88e9-561021e3dc9c)

and results in 
```
DEBUG   events  operation error IAM: PutRolePolicy, https response error StatusCode: 400, RequestID: XXXXXXX, MalformedPolicyDocument: Syntax errors in policy.
```

I tried using kubebuilder validation instead.  Ex.  `// +kubebuilder:validation:Format:=json`
But json is not in the type list that it can validate.
https://github.com/kubernetes/apiextensions-apiserver/blob/d323d00fb3d4c0c1532ac4fdbeec5aef190104ca/pkg/apiserver/validation/formats.go#L26C1-L51

I agree it would be best to be able to reject invalid policies as early in the process as possible and at the k8s API level.  Validation Webhook would be ideal.
I wonder if we could use something like this -> https://github.com/crossplane/crossplane/issues/3991
In the same way as validation webhooks can be used to validate compositions, I am wondering if we can use something similar for non-compositions?
There are many other instances where we want more validation than kubegen will give us. 

In the least it would be good to fail in the controller instead of passing malformed policies to AWS. 

For now - can we start with string and for enhancements:
1. Possibly add validation in the controllers (before ANY policies are sent to AWS API)
2. Add enhancement to also support `ResourcePolicy` - and prefer that over `string` if defined
3. Look at webhook validation or any other k8s validation which could be used

This PR is no worse than IAM Policies are handled elsewhere in `provider-aws`
It would be nice to get this finally merged, and it would be nice to have a solution which could be used across any aws policies included in this provider